{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running on Apache Spark version 2.1.0\n",
      "SparkUI available at http://129.94.72.55:4041\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.1-c72c76d\n"
     ]
    }
   ],
   "source": [
    "from hail import *\n",
    "hc = HailContext(tmp_dir = '/nvme/tmp/shusson')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vds = hc.read('/nvme/tmp/shusson/sample_stats.vds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Summary(samples=2845, variants=84222, call_rate=0.989135, contigs=['X', '12', '8', '19', '4', '15', '11', '9', 'Y', '22', '13', '16', '5', '10', '21', '6', '1', '17', '14', 'MT', '20', '2', '18', '7', '3'], multiallelics=0, snps=66323, mnps=0, insertions=8225, deletions=9674, complex=0, star=0, max_alleles=2)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vds.summarize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if not vds.was_split():\n",
    "    vds = vds.split_multi()\n",
    "\n",
    "vds = vds.annotate_variants_expr('va.allele = va.alleles[va.aIndex-1]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "kt=vds.variants_table()\n",
    "\n",
    "kt = kt.annotate('VARIANT=v, \\\n",
    "CHROMOSOME=v.contig, \\\n",
    "START=v.start, \\\n",
    "REF=v.ref, \\\n",
    "ALT=v.alt, \\\n",
    "RSID=orElse(va.allele.annotation.rsid, \"\\\\\\N\"), \\\n",
    "AC=va.allele.metrics.allele_counts.total.toInt(), \\\n",
    "AF=if (isMissing(va.allele.metrics.allele_frequencies.total)) \"\\\\\\N\" else str(va.allele.metrics.allele_frequencies.total),\\\n",
    "nHomRef=va.extra.nHomRef.toInt(), \\\n",
    "nHet=va.extra.nHet.total.toInt(), \\\n",
    "nHomVar=va.extra.nHomVar.toInt(), \\\n",
    "TYPE=v.altAllele.category(), \\\n",
    "CATO=if (isMissing(va.allele.annotation.predictions.CATO.score)) \"\\\\\\N\" else str(va.allele.annotation.predictions.CATO.score), \\\n",
    "eigen=if (isMissing(va.allele.annotation.predictions.Eigen.EigenRaw)) \"\\\\\\N\" else str(va.allele.annotation.predictions.Eigen.EigenRaw), \\\n",
    "polyPhen=if (isMissing(va.allele.annotation.vep.predictions.PolyPhen)) \"\\\\\\N\" else str(va.allele.annotation.vep.predictions.PolyPhen), \\\n",
    "sift=if (isMissing(va.allele.annotation.vep.predictions.SIFT)) \"\\\\\\N\" else str(va.allele.annotation.vep.predictions.SIFT), \\\n",
    "HGVSc=if (isMissing(va.allele.annotation.vep.HGVS.HGVSc)) \"\\\\\\N\" else str(va.allele.annotation.vep.HGVS.HGVSc), \\\n",
    "HGVSp=if (isMissing(va.allele.annotation.vep.HGVS.HGVSp)) \"\\\\\\N\" else str(va.allele.annotation.vep.HGVS.HGVSp), \\\n",
    "tgpAF=if (isMissing(va.allele.annotation.freqs.tgp.AF)) \"\\\\\\N\" else str(va.allele.annotation.freqs.tgp.AF), \\\n",
    "hrcAF=if (isMissing(va.allele.annotation.freqs.hrc.AF)) \"\\\\\\N\" else str(va.allele.annotation.freqs.hrc.AF), \\\n",
    "gnomadAF=if (isMissing(va.allele.annotation.freqs.gnomad.AF)) \"\\\\\\N\" else str(va.allele.annotation.freqs.gnomad.AF), \\\n",
    "feature=orElse(va.allele.annotation.vep.consequences.feature.id, \"\\\\\\N\"), \\\n",
    "consequences=orElse(va.allele.annotation.vep.consequences.consequences.mkString(\",\"), \"\\\\\\N\"), \\\n",
    "gene=orElse(va.allele.annotation.vep.cross_references.Ensembl_gene, \"\\\\\\N\"), \\\n",
    "clinvar=orElse(va.allele.annotation.clinvar.ClinicalSignificance.mkString(\",\"), \"\\\\\\N\"), \\\n",
    "wasSplit=va.extra.wasSplit')\n",
    "\n",
    "kt = kt.select([u'VARIANT', u'CHROMOSOME', u'START', u'REF', u'ALT', u'RSID', u'AC', u'AF', u'nHomRef', u'nHet', u'nHomVar', u'TYPE', 'CATO', 'eigen', 'sift', 'HGVSc', 'HGVSp', 'polyPhen', 'tgpAF', 'hrcAF', 'gnomadAF', 'feature', 'consequences', 'gene','clinvar', 'wasSplit'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2017-10-25 13:10:03 Hail: INFO: while writing:\n",
      "    /nvme/tmp/shusson/mgrb_sample.tsv\n",
      "  merge time: 367.510ms\n"
     ]
    }
   ],
   "source": [
    "kt.export(\"/nvme/tmp/shusson/mgrb_sample.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "kt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Struct{\n",
      "     variant: Struct{\n",
      "         qual: Double,\n",
      "         filters: Set[String],\n",
      "         metrics: Struct{\n",
      "             sample_counts: Struct{\n",
      "                 female: Long,\n",
      "                 male: Long,\n",
      "                 total: Long\n",
      "             },\n",
      "             chromatid_counts: Struct{\n",
      "                 female: Double,\n",
      "                 male: Double,\n",
      "                 total: Double\n",
      "             },\n",
      "             gatk: Struct{\n",
      "                 AN: Int,\n",
      "                 BaseQRankSum: Double,\n",
      "                 ClippingRankSum: Double,\n",
      "                 DB: Boolean,\n",
      "                 DP: Int,\n",
      "                 DS: Boolean,\n",
      "                 END: Int,\n",
      "                 ExcessHet: Double,\n",
      "                 FS: Double,\n",
      "                 HaplotypeScore: Double,\n",
      "                 InbreedingCoeff: Double,\n",
      "                 MQ: Double,\n",
      "                 MQRankSum: Double,\n",
      "                 NEGATIVE_TRAIN_SITE: Boolean,\n",
      "                 POSITIVE_TRAIN_SITE: Boolean,\n",
      "                 QD: Double,\n",
      "                 RAW_MQ: Double,\n",
      "                 ReadPosRankSum: Double,\n",
      "                 SOR: Double,\n",
      "                 VQSLOD: Double,\n",
      "                 culprit: String\n",
      "             }\n",
      "         }\n",
      "     },\n",
      "     genotypes: Struct{\n",
      "         counts: Struct{\n",
      "             female: Array[Int],\n",
      "             male: Array[Int],\n",
      "             total: Array[Int]\n",
      "         }\n",
      "     },\n",
      "     alleles: Array[Struct{\n",
      "         metrics: Struct{\n",
      "             allele_counts: Struct{\n",
      "                 female: Double,\n",
      "                 male: Double,\n",
      "                 total: Double\n",
      "             },\n",
      "             allele_frequencies: Struct{\n",
      "                 female: Double,\n",
      "                 male: Double,\n",
      "                 total: Double\n",
      "             },\n",
      "             gatk: Struct{\n",
      "                 AC: Int,\n",
      "                 AF: Double,\n",
      "                 MLEAC: Int,\n",
      "                 MLEAF: Double\n",
      "             }\n",
      "         },\n",
      "         annotation: Struct{\n",
      "             rsid: String,\n",
      "             vep: Struct{\n",
      "                 consequences: Struct{\n",
      "                     feature: Struct{\n",
      "                         id: String,\n",
      "                         type: String,\n",
      "                         biotype: String,\n",
      "                         strand: Int,\n",
      "                         canonical: String\n",
      "                     },\n",
      "                     position_in_feature: Struct{\n",
      "                         cDNA: String,\n",
      "                         CDS: String,\n",
      "                         protein: String,\n",
      "                         exon: String,\n",
      "                         intron: String,\n",
      "                         overlapping_domains: Array[String]\n",
      "                     },\n",
      "                     variant_class: String,\n",
      "                     distance_from_feature: Int,\n",
      "                     consequences: Array[String],\n",
      "                     impact_class: String,\n",
      "                     codon_change: String,\n",
      "                     residue_change: String,\n",
      "                     motif_disruption: Struct{\n",
      "                         motif: String,\n",
      "                         motif_pos: Int,\n",
      "                         high_inf_pos: String,\n",
      "                         motif_score_change: Float\n",
      "                     }\n",
      "                 },\n",
      "                 predictions: Struct{\n",
      "                     ClinVar: Array[String],\n",
      "                     PolyPhen: String,\n",
      "                     SIFT: String\n",
      "                 },\n",
      "                 HGVS: Struct{\n",
      "                     HGVSc: String,\n",
      "                     HGVSp: String,\n",
      "                     HGVS_offset: Int\n",
      "                 },\n",
      "                 population_frequencies: Struct{\n",
      "                     TGP: Struct{\n",
      "                         AF: Float,\n",
      "                         AFR_AF: Float,\n",
      "                         AMR_AF: Float,\n",
      "                         EAS_AF: Float,\n",
      "                         EUR_AF: Float,\n",
      "                         SAS_AF: Float\n",
      "                     },\n",
      "                     ESP: Struct{\n",
      "                         AA_AF: Float,\n",
      "                         EA_AF: Float\n",
      "                     },\n",
      "                     GnomAD: Struct{\n",
      "                         AF: Float,\n",
      "                         AFR_AF: Float,\n",
      "                         AMR_AF: Float,\n",
      "                         ASJ_AF: Float,\n",
      "                         EAS_AF: Float,\n",
      "                         FIN_AF: Float,\n",
      "                         NFE_AF: Float,\n",
      "                         OTH_AF: Float,\n",
      "                         SAS_AF: Float\n",
      "                     },\n",
      "                     MAX_AF: Float,\n",
      "                     MAX_AF_POPS: Array[String]\n",
      "                 },\n",
      "                 cross_references: Struct{\n",
      "                     Ensembl_gene: String,\n",
      "                     Ensembl_protein: String,\n",
      "                     PUBMED: Array[Int],\n",
      "                     CCDS: String,\n",
      "                     SWISSPROT: String,\n",
      "                     UNIPARC: String,\n",
      "                     symbol: String,\n",
      "                     symbol_source: String,\n",
      "                     HGNC: Int,\n",
      "                     existing_variation: String\n",
      "                 }\n",
      "             },\n",
      "             clinvar: Struct{\n",
      "                 AlleleID: Int,\n",
      "                 ClinicalSignificance: Array[String],\n",
      "                 ClinSigSimple: Int,\n",
      "                 rsid: Int,\n",
      "                 ReviewStatus: String,\n",
      "                 NumberSubmitters: Int\n",
      "             },\n",
      "             predictions: Struct{\n",
      "                 CATO: Struct{\n",
      "                     score: Double,\n",
      "                     motif: String,\n",
      "                     strand: Int,\n",
      "                     motifpos: Int,\n",
      "                     celltypes: Array[String]\n",
      "                 },\n",
      "                 Eigen: Struct{\n",
      "                     EigenRaw: Double,\n",
      "                     EigenPhred: Double,\n",
      "                     EigenPCRaw: Double,\n",
      "                     EigenPCPhred: Double\n",
      "                 }\n",
      "             },\n",
      "             freqs: Struct{\n",
      "                 tgp: Struct{\n",
      "                     AC: Int,\n",
      "                     AF: Double,\n",
      "                     NS: Int,\n",
      "                     AN: Int,\n",
      "                     AF_EAS: Double,\n",
      "                     AF_EUR: Double,\n",
      "                     AF_AFR: Double,\n",
      "                     AF_AMR: Double,\n",
      "                     AF_SAS: Double,\n",
      "                     AC_Hail: Int,\n",
      "                     AF_Hail: Double,\n",
      "                     NS_Hail: Int\n",
      "                 },\n",
      "                 hrc: Struct{\n",
      "                     AC: Int,\n",
      "                     AN: Int,\n",
      "                     AF: Double,\n",
      "                     AC_EXCLUDING_1000G: Int,\n",
      "                     AN_EXCLUDING_1000G: Int,\n",
      "                     AF_EXCLUDING_1000G: Double\n",
      "                 },\n",
      "                 gnomad: Struct{\n",
      "                     AC: Int,\n",
      "                     AF: Double,\n",
      "                     AN: Int,\n",
      "                     AC_AFR: Int,\n",
      "                     AC_AMR: Int,\n",
      "                     AC_ASJ: Int,\n",
      "                     AC_EAS: Int,\n",
      "                     AC_FIN: Int,\n",
      "                     AC_NFE: Int,\n",
      "                     AC_OTH: Int,\n",
      "                     AN_AFR: Int,\n",
      "                     AN_AMR: Int,\n",
      "                     AN_ASJ: Int,\n",
      "                     AN_EAS: Int,\n",
      "                     AN_FIN: Int,\n",
      "                     AN_NFE: Int,\n",
      "                     AN_OTH: Int,\n",
      "                     AF_AFR: Double,\n",
      "                     AF_AMR: Double,\n",
      "                     AF_ASJ: Double,\n",
      "                     AF_EAS: Double,\n",
      "                     AF_FIN: Double,\n",
      "                     AF_NFE: Double,\n",
      "                     AF_OTH: Double\n",
      "                 }\n",
      "             }\n",
      "         }\n",
      "     }],\n",
      "     locus: Struct{\n",
      "         badCoverage_dist: Int,\n",
      "         badCoverage: Boolean,\n",
      "         badComplexity_dist: Int,\n",
      "         badComplexity: Boolean,\n",
      "         badRepeat_dist: Int,\n",
      "         badRepeat: Boolean,\n",
      "         badMappability_dist: Int,\n",
      "         badMappability: Boolean,\n",
      "         badEncodeExcluded_dist: Int,\n",
      "         badEncodeExcluded: Boolean,\n",
      "         goodGiaBHighConf_dist: Int,\n",
      "         goodGiaBHighConf: Boolean,\n",
      "         badPAR: Boolean,\n",
      "         tier: Int\n",
      "     },\n",
      "     aIndex: Int,\n",
      "     wasSplit: Boolean,\n",
      "     info: Struct{\n",
      "         nHomVarFemale: Long,\n",
      "         nHomVarMale: Long,\n",
      "         nHomRefFemale: Long,\n",
      "         nHomRefMale: Long,\n",
      "         nHetFemale: Long,\n",
      "         nHetMale: Long,\n",
      "         nHomVar: Long,\n",
      "         nHomRef: Long,\n",
      "         nHet: Long\n",
      "     },\n",
      "     fred: Struct{\n",
      "         gtCounts: Struct{\n",
      "             female: Array[Int],\n",
      "             male: Array[Int],\n",
      "             total: Array[Int]\n",
      "         },\n",
      "         alleleCounts: Struct{\n",
      "             female: Array[Double],\n",
      "             male: Array[Double],\n",
      "             total: Array[Double]\n",
      "         },\n",
      "         AC: Array[Double],\n",
      "         AF: Array[Double]\n",
      "     },\n",
      "     allele: Struct{\n",
      "         metrics: Struct{\n",
      "             allele_counts: Struct{\n",
      "                 female: Double,\n",
      "                 male: Double,\n",
      "                 total: Double\n",
      "             },\n",
      "             allele_frequencies: Struct{\n",
      "                 female: Double,\n",
      "                 male: Double,\n",
      "                 total: Double\n",
      "             },\n",
      "             gatk: Struct{\n",
      "                 AC: Int,\n",
      "                 AF: Double,\n",
      "                 MLEAC: Int,\n",
      "                 MLEAF: Double\n",
      "             }\n",
      "         },\n",
      "         annotation: Struct{\n",
      "             rsid: String,\n",
      "             vep: Struct{\n",
      "                 consequences: Struct{\n",
      "                     feature: Struct{\n",
      "                         id: String,\n",
      "                         type: String,\n",
      "                         biotype: String,\n",
      "                         strand: Int,\n",
      "                         canonical: String\n",
      "                     },\n",
      "                     position_in_feature: Struct{\n",
      "                         cDNA: String,\n",
      "                         CDS: String,\n",
      "                         protein: String,\n",
      "                         exon: String,\n",
      "                         intron: String,\n",
      "                         overlapping_domains: Array[String]\n",
      "                     },\n",
      "                     variant_class: String,\n",
      "                     distance_from_feature: Int,\n",
      "                     consequences: Array[String],\n",
      "                     impact_class: String,\n",
      "                     codon_change: String,\n",
      "                     residue_change: String,\n",
      "                     motif_disruption: Struct{\n",
      "                         motif: String,\n",
      "                         motif_pos: Int,\n",
      "                         high_inf_pos: String,\n",
      "                         motif_score_change: Float\n",
      "                     }\n",
      "                 },\n",
      "                 predictions: Struct{\n",
      "                     ClinVar: Array[String],\n",
      "                     PolyPhen: String,\n",
      "                     SIFT: String\n",
      "                 },\n",
      "                 HGVS: Struct{\n",
      "                     HGVSc: String,\n",
      "                     HGVSp: String,\n",
      "                     HGVS_offset: Int\n",
      "                 },\n",
      "                 population_frequencies: Struct{\n",
      "                     TGP: Struct{\n",
      "                         AF: Float,\n",
      "                         AFR_AF: Float,\n",
      "                         AMR_AF: Float,\n",
      "                         EAS_AF: Float,\n",
      "                         EUR_AF: Float,\n",
      "                         SAS_AF: Float\n",
      "                     },\n",
      "                     ESP: Struct{\n",
      "                         AA_AF: Float,\n",
      "                         EA_AF: Float\n",
      "                     },\n",
      "                     GnomAD: Struct{\n",
      "                         AF: Float,\n",
      "                         AFR_AF: Float,\n",
      "                         AMR_AF: Float,\n",
      "                         ASJ_AF: Float,\n",
      "                         EAS_AF: Float,\n",
      "                         FIN_AF: Float,\n",
      "                         NFE_AF: Float,\n",
      "                         OTH_AF: Float,\n",
      "                         SAS_AF: Float\n",
      "                     },\n",
      "                     MAX_AF: Float,\n",
      "                     MAX_AF_POPS: Array[String]\n",
      "                 },\n",
      "                 cross_references: Struct{\n",
      "                     Ensembl_gene: String,\n",
      "                     Ensembl_protein: String,\n",
      "                     PUBMED: Array[Int],\n",
      "                     CCDS: String,\n",
      "                     SWISSPROT: String,\n",
      "                     UNIPARC: String,\n",
      "                     symbol: String,\n",
      "                     symbol_source: String,\n",
      "                     HGNC: Int,\n",
      "                     existing_variation: String\n",
      "                 }\n",
      "             },\n",
      "             clinvar: Struct{\n",
      "                 AlleleID: Int,\n",
      "                 ClinicalSignificance: Array[String],\n",
      "                 ClinSigSimple: Int,\n",
      "                 rsid: Int,\n",
      "                 ReviewStatus: String,\n",
      "                 NumberSubmitters: Int\n",
      "             },\n",
      "             predictions: Struct{\n",
      "                 CATO: Struct{\n",
      "                     score: Double,\n",
      "                     motif: String,\n",
      "                     strand: Int,\n",
      "                     motifpos: Int,\n",
      "                     celltypes: Array[String]\n",
      "                 },\n",
      "                 Eigen: Struct{\n",
      "                     EigenRaw: Double,\n",
      "                     EigenPhred: Double,\n",
      "                     EigenPCRaw: Double,\n",
      "                     EigenPCPhred: Double\n",
      "                 }\n",
      "             },\n",
      "             freqs: Struct{\n",
      "                 tgp: Struct{\n",
      "                     AC: Int,\n",
      "                     AF: Double,\n",
      "                     NS: Int,\n",
      "                     AN: Int,\n",
      "                     AF_EAS: Double,\n",
      "                     AF_EUR: Double,\n",
      "                     AF_AFR: Double,\n",
      "                     AF_AMR: Double,\n",
      "                     AF_SAS: Double,\n",
      "                     AC_Hail: Int,\n",
      "                     AF_Hail: Double,\n",
      "                     NS_Hail: Int\n",
      "                 },\n",
      "                 hrc: Struct{\n",
      "                     AC: Int,\n",
      "                     AN: Int,\n",
      "                     AF: Double,\n",
      "                     AC_EXCLUDING_1000G: Int,\n",
      "                     AN_EXCLUDING_1000G: Int,\n",
      "                     AF_EXCLUDING_1000G: Double\n",
      "                 },\n",
      "                 gnomad: Struct{\n",
      "                     AC: Int,\n",
      "                     AF: Double,\n",
      "                     AN: Int,\n",
      "                     AC_AFR: Int,\n",
      "                     AC_AMR: Int,\n",
      "                     AC_ASJ: Int,\n",
      "                     AC_EAS: Int,\n",
      "                     AC_FIN: Int,\n",
      "                     AC_NFE: Int,\n",
      "                     AC_OTH: Int,\n",
      "                     AN_AFR: Int,\n",
      "                     AN_AMR: Int,\n",
      "                     AN_ASJ: Int,\n",
      "                     AN_EAS: Int,\n",
      "                     AN_FIN: Int,\n",
      "                     AN_NFE: Int,\n",
      "                     AN_OTH: Int,\n",
      "                     AF_AFR: Double,\n",
      "                     AF_AMR: Double,\n",
      "                     AF_ASJ: Double,\n",
      "                     AF_EAS: Double,\n",
      "                     AF_FIN: Double,\n",
      "                     AF_NFE: Double,\n",
      "                     AF_OTH: Double\n",
      "                 }\n",
      "             }\n",
      "         }\n",
      "     }\n",
      " }\n"
     ]
    }
   ],
   "source": [
    "from pprint import pprint\n",
    "pprint(vds.variant_schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
